### PR TITLE
Add `inspec env` command to configure shell tab-completion

### LINKED ISSUE #607: tab-completion support in the inspec command line and in the inspec shell :)
Tab-completion ... at least limited support may be good for the release.

I will see if I can grok bash-complete and add it in.

### PR DESCRIPTION
**Pull request by stevendanna**
This adds a new subcommand:

```
inspec env [SHELL]
```

which outputs a shell-appropriate completion script that the user can
source into their shell:

```
eval "$(inspec env SHELL)"
```

Currently, we provide completions for ZSH and Bash. The completion
scripts are generated from the data Thor collects.

If the user doesn't provide SHELL we attempt to detect what the user's
shell may be using a number of methods.

Fixes #607

Signed-off-by: Steven Danna steve@chef.io

### CODE REVIEW

File: lib/inspec/shell_detector.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    def shell!
+      @shell = detect
+    end
+
+    private
+
+    def detect
+      # Most of our detection code assumes a unix-like environment
+      return nil if (RbConfig::CONFIG['host_os'] =~ /mswin|mingw|cygwin/)
```
Comment (chris-rock) [inline lib/inspec/shell_detector.rb]: May it make sense to output a warning? Otherwise the user may be confused that nothing has been executed?

**Comment by chris-rock** [comment]
@stevendanna great work. Bash completion works like a charm. I've some troubles with zsh and get the following errors:

```
eval "$(inspec env zsh)"
zsh: parse error near `>'
```

My generated content is available here: https://gist.github.com/chris-rock/a30592b0b82d13136772c10162dc5752

File: lib/inspec/env_printer.rb [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      $stderr.puts "# No completion script for #{@shell}!"
+      print_detection_warning($stderr) if @detected
+      exit 1
+    end
+
+    def exit_no_shell
+      if @detected
+        $stderr.puts '# Unable to automatically detect shell and no shell was provided.'
+      end
+      $stderr.puts <<EOF
```
Comment (chris-rock) [inline lib/inspec/env_printer.rb:88]: cool!

File: lib/inspec/env_printer.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+# To use this, eval it in your shell
+#
+#    eval "$(inspec env #{@shell})"
+#
+#
+EOF
+    end
+
+    def print_detection_warning(device)
+      device.puts <<EOF
```
Comment (chris-rock) [inline lib/inspec/env_printer.rb:66]: Its great that you can manually override the detection

File: lib/inspec/shell_detector.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+      shellpath = detect_by_ppid
```
Comment (chris-rock) [inline lib/inspec/shell_detector.rb:41]: nice to see the different mechanisms! awesome work @stevendanna

**Comment by chris-rock** [comment]
Awesome implementation @stevendanna LGTM once the zsh is working

**Comment by stevendanna** [comment]
@chris-rock I think ZSH should be working now.  The issue was we weren't escaping the description text and you had plugins that had commands with `'` in their description.  Now we use `shellwords` to escape any special characters in the description.  Let me know if it works for you.

**Comment by chris-rock** [comment]
@stevendanna works like a charm now!

**Comment by stevendanna** [comment]
Merging this despite failing tests since TravisCI seems to be having an incident of some sort.

**Pull request merged**